Handle frameworks and binding packages in Hot Restart

mattjohnsonpint opened the pull request:
- Supports either regular framework or xcframework files coming from nuget packages
- Updates the `Unzip` task to properly create symbolic links, which may occur within resource zips containing xcframeworks

Fixes #16571

@rolfbjarne

mattjohnsonpint commented:
@microsoft-github-policy-service agree

rolfbjarne commented on msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.HotRestart.targets:
```diff
@@ -86,7 +87,34 @@
 
-	<Target Name="_CollectHotRestartFrameworks" >
+	<!-- Also include frameworks from packages -->
+	<Target Name="_CollectHotRestartFrameworksFromPackages"
```
Nothing depends on this target, so as far as I can tell it's never called?

mattjohnsonpint replied:
Oops - edit fail.  Fixed.

rolfbjarne commented on msbuild/Xamarin.iOS.Tasks.Windows/Zip.cs:
```diff
@@ -35,6 +35,56 @@ static void CopyDirectory (string source, string destination)
 			}
 		}
+
+		private static void ExtractWithSymlinksToDirectory (string sourceFileName, string destinationDirectoryName)
```
Same: nothing calls this method.

mattjohnsonpint replied:
Oops - edit fail.  Fixed.

dalexsoto commented:
/azp run

dalexsoto commented:
/azp run

vs-mobiletools-engineering-service2 commented:
# :fire: [PR Build] Build failed :fire:

Build failed for the job 'Build packages'

[Pipeline](https://dev.azure.com/devdiv/DevDiv/_build/index?buildId=6935270&view=ms.vss-test-web.test-result-details) on Agent 
Hash: [](https://github.com/xamarin/xamarin-macios/pull/16605/commits/)  [PR build]

[comment]: <> (This is a comment added by Azure DevOps)

dalexsoto commented:
Looks like there is a build error going on

```
Build FAILED.

       "/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks.sln" (default target) (1:2) ->
       "/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.Tasks.Windows.csproj" (default target) (4:6) ->
       (CoreCompile target) -> 
         /Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Zip.cs(55,37): error CS1061: 'FileInfo' does not contain a definition for 'LinkTarget' and no accessible extension method 'LinkTarget' accepting a first argument of type 'FileInfo' could be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.Tasks.Windows.csproj]
         /Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Zip.cs(68,15): error CS0117: 'Directory' does not contain a definition for 'CreateSymbolicLink' [/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.Tasks.Windows.csproj]
         /Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Zip.cs(76,11): error CS0117: 'File' does not contain a definition for 'CreateSymbolicLink' [/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.Tasks.Windows.csproj]
         /Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Zip.cs(100,73): error CS1061: 'ZipArchiveEntry' does not contain a definition for 'ExternalAttributes' and no accessible extension method 'ExternalAttributes' accepting a first argument of type 'ZipArchiveEntry' could be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/4/s/xamarin-macios/msbuild/Xamarin.iOS.Tasks.Windows/Xamarin.iOS.Tasks.Windows.csproj]

    4 Warning(s)
    4 Error(s)
```

mattjohnsonpint commented:
Am I not allowed to use .NET 6 APIs?

mattjohnsonpint commented:
I added a `net6.0` target, and updated the code to only extract symlinks for net6+.  Is that allowed? Will the net6 version make its way to the correct package to be used by VS at runtime?  I'm not set up to fully build and test end-to-end, so I'm flying blind a bit here.  I'd appreciate any assistance.

BTW - The other approach I considered would be to not rely on these new symlink APIs but instead to re-implement the `readlink` algorithm in C# to work directly against the zip archive.  The result would be files on disk are always fully materialized, even if there's multiple copies of them.  It is much more difficult than it seems though, as links can point to links, or contain links as portions of the path.

rolfbjarne commented:
So there are a few problems here:

> I added a `net6.0` target, and updated the code to only extract symlinks for net6+. Is that allowed?

Yes, it's allowed, but...

> Will the net6 version make its way to the correct package to be used by VS at runtime?

No, not at the moment. We'll eventually get there, but we're not there yet.

> BTW - The other approach I considered would be to not rely on these new symlink APIs but instead to re-implement the `readlink` algorithm in C# to work directly against the zip archive. The result would be files on disk are always fully materialized, even if there's multiple copies of them. It is much more difficult than it seems though, as links can point to links, or contain links as portions of the path.

Unfortunately there are a few problems here:

* I believe Windows may require admin privileges to create symlinks in some cases.
* I don't think the code that installs the app on the device knows about windows symlinks.

On the other hand I realize that zipped xcframeworks might contain symlinks that won't be needed (if they're for a different platform), so it certainly makes sense to find a solution here.

However, given that implementing and testing this solution is likely to take a while, my suggestion would be to split this PR in two: one PR that only adds Hot Restart support for (xc)frameworks and binding resource packages, and then another one that tackles the symlink problem.

From what you've done in this PR it looks like the first part would only contain changes to .props and .targets files, so that's much easier to test yourself (you can just change these files locally).

mattjohnsonpint commented:
OK, I will update this tomorrow and remove the zip changes.

Indeed in my own library, we only have symlink in the maccatalyst target.  The ios target has none, and the zip extract finds them just fine.  Since only the ios target is needed for hot restart, I think it is fine to not support symlinks for now.

rolfbjarne commented:
Thanks! I'll have a look and see if I can write tests for this (and update the PR accordingly).

mattjohnsonpint commented:
Found and fixed a small but important bug with an extra trailing backslash, which was throwing off some subsequent checks such as creating the correct stamp file in the DynamicFrameworks folder.

rolfbjarne commented:
/azp run

vs-mobiletools-engineering-service2 commented:
# :books: [PR Build] Artifacts :books:

## Packages generated

<details><summary>View packages</summary>

* [Microsoft.iOS.Bundle.16.1.586-ci.pr.gh16605.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Bundle.16.1.586-ci.pr.gh16605.pkg)
* [Microsoft.iOS.Ref.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Ref.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Runtime.ios-arm.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Runtime.ios-arm.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Runtime.ios-arm64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Runtime.ios-arm64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Runtime.iossimulator-arm64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Runtime.iossimulator-arm64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Runtime.iossimulator-x64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Runtime.iossimulator-x64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Runtime.iossimulator-x86.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Runtime.iossimulator-x86.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Sdk.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Sdk.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Templates.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Templates.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.iOS.Windows.Sdk.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.iOS.Windows.Sdk.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.MacCatalyst.Bundle.16.1.586-ci.pr.gh16605.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Bundle.16.1.586-ci.pr.gh16605.pkg)
* [Microsoft.MacCatalyst.Ref.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Ref.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.MacCatalyst.Runtime.maccatalyst-arm64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Runtime.maccatalyst-arm64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.MacCatalyst.Runtime.maccatalyst-x64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Runtime.maccatalyst-x64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.MacCatalyst.Sdk.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Sdk.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.MacCatalyst.Templates.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.MacCatalyst.Templates.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.macOS.Bundle.13.0.1141-ci.pr.gh16605.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Bundle.13.0.1141-ci.pr.gh16605.pkg)
* [Microsoft.macOS.Ref.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Ref.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.macOS.Runtime.osx-arm64.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Runtime.osx-arm64.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.macOS.Runtime.osx-x64.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Runtime.osx-x64.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.macOS.Sdk.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Sdk.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.macOS.Templates.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.macOS.Templates.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.NET.Sdk.iOS.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.NET.Sdk.iOS.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.NET.Sdk.MacCatalyst.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.NET.Sdk.MacCatalyst.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.NET.Sdk.macOS.Manifest-7.0.100.13.0.1141-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.NET.Sdk.macOS.Manifest-7.0.100.13.0.1141-ci.pr.gh16605.nupkg)
* [Microsoft.NET.Sdk.tvOS.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.NET.Sdk.tvOS.Manifest-7.0.100.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.NET.Workload.iOS.16.1.586-ci.pr.gh16605.msi](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.NET.Workload.iOS.16.1.586-ci.pr.gh16605.msi)
* [Microsoft.tvOS.Bundle.16.1.586-ci.pr.gh16605.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Bundle.16.1.586-ci.pr.gh16605.pkg)
* [Microsoft.tvOS.Ref.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Ref.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.tvOS.Runtime.tvos-arm64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Runtime.tvos-arm64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.tvOS.Runtime.tvossimulator-arm64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Runtime.tvossimulator-arm64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.tvOS.Runtime.tvossimulator-x64.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Runtime.tvossimulator-x64.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.tvOS.Sdk.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Sdk.16.1.586-ci.pr.gh16605.nupkg)
* [Microsoft.tvOS.Templates.16.1.586-ci.pr.gh16605.nupkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/Microsoft.tvOS.Templates.16.1.586-ci.pr.gh16605.nupkg)
* [xamarin.ios-16.1.1.459.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/xamarin.ios-16.1.1.459.pkg)
* [xamarin.mac-9.0.0.459.pkg](https://dl.internalx.com/wrench/merge/6805d6775f0c396eb5240058d890b71e9fbdf3d7/6965740/package/xamarin.mac-9.0.0.459.pkg)
</details>

[Pipeline](https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=6965740&view=ms.vss-test-web.test-result-details) on Agent XAMBOT-1100.Monterey'
Hash: [b5e6db47a305c26f97208c4722b79c0dbb025f16](https://github.com/xamarin/xamarin-macios/pull/16605/commits/b5e6db47a305c26f97208c4722b79c0dbb025f16)  [PR build]

[comment]: <> (This is a comment added by Azure DevOps)

vs-mobiletools-engineering-service2 commented:
# :computer: [PR Build] Tests on macOS M1 - Mac Big Sur (11.5) passed :computer:

:white_check_mark: **All** tests on macOS M1 - Mac Big Sur (11.5) passed.

[Pipeline](https://dev.azure.com/devdiv/DevDiv/_build/index?buildId=6965740&view=ms.vss-test-web.test-result-details) on Agent 
Hash: [b5e6db47a305c26f97208c4722b79c0dbb025f16](https://github.com/xamarin/xamarin-macios/pull/16605/commits/b5e6db47a305c26f97208c4722b79c0dbb025f16)  [PR build]

[comment]: <> (This is a comment added by Azure DevOps)

vs-mobiletools-engineering-service2 commented:
# :rocket: [CI Build] Test results :rocket:

# Test results
:white_check_mark: All tests passed on VSTS: simulator tests.

:tada: All 223 tests passed :tada:

## Tests counts
:white_check_mark: bcl: All 69 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_bcl-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_bcl-1&api-version=6.0&$format=zip)
:white_check_mark: cecil: All 1 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_cecil-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_cecil-1&api-version=6.0&$format=zip)
:white_check_mark: dotnettests: All 1 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_dotnettests-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_dotnettests-1&api-version=6.0&$format=zip)
:white_check_mark: fsharp: All 7 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_fsharp-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_fsharp-1&api-version=6.0&$format=zip)
:white_check_mark: framework: All 8 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_framework-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_framework-1&api-version=6.0&$format=zip)
:white_check_mark: generator: All 2 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_generator-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_generator-1&api-version=6.0&$format=zip)
:white_check_mark: interdependent_binding_projects: All 7 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_interdependent_binding_projects-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_interdependent_binding_projects-1&api-version=6.0&$format=zip)
:white_check_mark: install_source: All 1 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_install_source-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_install_source-1&api-version=6.0&$format=zip)
:white_check_mark: introspection: All 8 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_introspection-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_introspection-1&api-version=6.0&$format=zip)
:white_check_mark: linker: All 65 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_linker-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_linker-1&api-version=6.0&$format=zip)
:white_check_mark: mac_binding_project: All 1 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_mac_binding_project-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_mac_binding_project-1&api-version=6.0&$format=zip)
:white_check_mark: mmp: All 2 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_mmp-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_mmp-1&api-version=6.0&$format=zip)
:white_check_mark: mononative: All 12 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_mononative-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_mononative-1&api-version=6.0&$format=zip)
:white_check_mark: monotouch: All 23 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_monotouch-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_monotouch-1&api-version=6.0&$format=zip)
:white_check_mark: msbuild: All 2 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_msbuild-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_msbuild-1&api-version=6.0&$format=zip)
:white_check_mark: mtouch: All 1 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_mtouch-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_mtouch-1&api-version=6.0&$format=zip)
:white_check_mark: xammac: All 3 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_xammac-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_xammac-1&api-version=6.0&$format=zip)
:white_check_mark: xcframework: All 8 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_xcframework-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_xcframework-1&api-version=6.0&$format=zip)
:white_check_mark: xtro: All 2 tests passed. [Html Report (VSDrops)](https://vsdrop.corp.microsoft.com/file/v1/xamarin-macios/device-tests/20221115.8/6965740//simulator_xtro-1/;/tests/vsdrops_index.html) [Download](https://devdiv.visualstudio.com/DevDiv/_apis/build/builds/6965740/artifacts?artifactName=HtmlReport-simulator_xtro-1&api-version=6.0&$format=zip)

[comment]: <> (This is a test result report added by Azure DevOps)

[Pipeline](https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=6965740&view=ms.vss-test-web.test-result-details) on Agent 
Hash: [$(GIT_HASH)](https://github.com/xamarin/xamarin-macios/pull/16605/commits/$(GIT_HASH))  [PR build]

[comment]: <> (This is a comment added by Azure DevOps)

mandel-macaque commented:
@rolfbjarne can you take a second view on this.

rolfbjarne commented:
> @rolfbjarne can you take a second view on this.

Yes, I'm working on adding tests for this scenario.

dalexsoto reviewed (approved):
Awesome work! Once get the tests up we'll merge it! Thanks a lot @mattjohnsonpint ❤️

mattjohnsonpint commented:
Thanks!  Now if we can just #16001 fixed, we'll have all the iOS on Windows stuff working well. 🙂

rolfbjarne commented:
> Thanks! Now if we can just #16001 fixed, we'll have all the iOS on Windows stuff working well. 🙂

Yep, I'm working on that one together with adding a test case for this issue.

rolfbjarne commented:
So I wrote a test for this scenario, and it revealed that this wasn't as simple as we first thought - in fact it was _much_ more complicated.

The good news is that I fixed all those issues, and the corresponding PR as been merged, so this PR is no longer needed (this means that xcframeworks and binding projects will finally work correctly with Hot Restart - but since the changes required were quite invasive, they won't be shipped until .NET 8. If you're willing to try a preview, they'll be available in the upcoming .NET 8 Preview 4).
